### PR TITLE
feat(groups):  Added groups app and logic

### PR DESCRIPTION
**Pull request by Her0elt**
Feat(groups):  Added groups app and logic
Added a groups app and started the logic for groups with model, serializer, and view.
I also moved premissions and enums to util, and fixed imports. 

Comments/issues:

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] PR name is descriptive
- [ ] The PR includes the link to the relevant Trello task
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)

- [ ] `python manage.py runserver` was run locally witout errors

## Pull request type

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### CODE REVIEW

File: app/content/admin/admin.py [hunk 1]
```diff
@@ -14,6 +16,10 @@
 admin.site.register(models.UserChallenge)
 
 
+
+
+
+
```
Comment (Eirsteir) [inline app/content/admin/admin.py:22]: Remember to register the Group model here 😉

File: app/group/factories/group_factory.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+from factory.django import DjangoModelFactory
+
+from ..models import Group
+
+
+class GroupFactory(DjangoModelFactory):
+    class Meta:
+        model = Group
+        
+    slug = "test_group"
```
Comment (Eirsteir) [inline app/group/factories/group_factory.py:10]: This will case an IntegrityError because of a duplicate value for slug. You can solve this issue by using some of factory-boy's mechanisms for fields, such as `lazy_attribute `and `Iterator` or `Sequence`. 

Have a look at these resources: 
https://joequery.me/code/factory-boy-handle-unique-constraints/
https://factoryboy.readthedocs.io/en/stable/reference.html#sequence

File: app/group/factories/group_factory.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    description = "this is a test group"
+    contact = "test@test.com"
+    created = 2019
+    type = 1
```
Comment (Eirsteir) [inline app/group/factories/group_factory.py:14]: A big point of a Factory is that it generates random, but consistent data. This means that it will generate random data for each instance, but you can always assert that they are what you expect. 

You should take a look at the official documentation https://factoryboy.readthedocs.io/en/stable/introduction.html 

For instance, the description might be 
` description = factory.Faker("sentence", nb_words=100, variable_nb_words=True)`
The type might be an `Iterable`, choosing a random type from the `GroupTypeEnum`:
`type = factory.Iterator(GroupTypeEnum)` (not sure about the syntax here, you might have to add the argument `getter=lambda c: c[?]`

The bottom line is that the tests should replicate real data as much as possible so that we can catch any bugs in the values a user might throw at us 😉

File: app/group/models/group.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+from django.contrib.auth.models import Group as DjangoGroup
+from app.util.models import OptionalImage
+from enumchoicefield import EnumChoiceField
+
+
+class Group(DjangoGroup,OptionalImage):
+    slug = models.SlugField(max_length=50, primary_key=True)
+    description = models.TextField(max_length=1000, null=True, blank = True)
+    contact = models.EmailField(max_length=200, null=True, blank = True)
+    created = models.IntegerField(null=True, blank = True)
```
Comment (Eirsteir) [inline app/group/models/group.py:12]: Should this be a `DateTimeField`?

File: app/group/models/group.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    type = EnumChoiceField(GroupType)
```
Comment (Eirsteir) [inline app/group/models/group.py:13]: Can I propose something here? Add a `default=GroupType.OTHER`. 

This way we avoid the need to specify further without having to suffer from null-values in the data and information.

File: app/group/tests/test_group_model.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+from django.core.exceptions import ValidationError
+
+import pytest
+
+from ..factories import GroupFactory
+
+@pytest.fixture()
+def group():
+    return GroupFactory()
```
Comment (Eirsteir) [inline app/group/tests/test_group_model.py:9]: Fantastic! Before you jump into writing your moddel tests, I would like to remind you that you do not have to test anything about the model, _except_ from your custom logic (that includes __str__)😄 

You can read more here, under 'Testing' https://docs.google.com/document/d/1ZVGBfNAlDfP8TQxVbb2VSnql0rVZs3j6YQTYc3000pA/edit?usp=sharing

File: app/group/views/group.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    permission_classes = [IsHS]
+    queryset = Group.objects.all()
+    pagination_class = BasePagination
+
+    def get_permissions(self):
+        if self.request.method in ["GET"]:
+            self.permission_classes = [IsMember]
+        return super(GroupViewSet, self).get_permissions()
+
+    def retrieve(self, request, pk):
```
Comment (Eirsteir) [inline app/group/views/group.py:23]: You can change `pk` to `slug` if you define a lookup field on the ViewSet itself

File: app/group/views/group.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+from rest_framework import filters, viewsets
+from rest_framework.response import Response
+
+from ..models import Group
+from app.util.pagination import BasePagination
+from app.util.permissions import IsMember, IsHS
+from ..serializers import GroupSerializer
+
+
+class GroupViewSet(viewsets.ModelViewSet):
```
Comment (Eirsteir) [inline app/group/views/group.py:12]: I have to say, this ViewSet is very clean! 😍

File: app/group/views/group.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    pagination_class = BasePagination
+
+    def get_permissions(self):
+        if self.request.method in ["GET"]:
+            self.permission_classes = [IsMember]
+        return super(GroupViewSet, self).get_permissions()
+
+    def retrieve(self, request, pk):
+        try:
+            group = Group.objects.get(slug = pk)
```
Comment (Eirsteir) [inline app/group/views/group.py:25]: @Zenjjim what did we conclude with about using `get_or_create(slug=slug)` again?

File: app/group/views/group.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+            serializer = GroupSerializer(group, data=request.data, partial=True)
+            if serializer.is_valid():
+                serializer.save()
+                return Response({"detail": _("Group created")}, status=201)
+        except Group.DoesNotExist:
+            return Response({"detail": _("Group not found.")}, status=404)
+
+    def create(self, request):
+        try:
+            if Group.objects.filter(slug = request.data["name"]).exists():
```
Comment (Eirsteir) [inline app/group/views/group.py:45]: This can be simplified by using `Group.get_or_create(slug=slug)` and returning the group found/created

File: docker-compose.yml [hunk 11]
```diff
@@ -6,7 +6,7 @@ services:
     container_name: db
     restart: always
     ports:
-      - "3306:3306"
+      - "3305:3305"
```
Comment (Eirsteir) [inline docker-compose.yml:9]: Why did you change this?

**Comment by Eirsteir** [comment]
I think this is great Hermann! This code is clean, understandable and simple (that is a compliment) 🤩  The comments are mostly tips and best practices, so don't worry. 

I do however not agree with the new location of permissions.py in utils. These classes are an essential part of the application and not "helpers", but I understand why you have put them there. 

What I suggest is to move them and enums.py to the top level directory (app) and plan for a future where the common permissions and enums might reside in a "common" app or a "permissions" app. Or simply keep commons in the top level directory and create new files in the sub-apps.

File: docker-compose.yml (hunk 11 above)
Comment (Her0elt) [reply docker-compose.yml:9]: im going to change it back, my mysql server runs on 3306 so i just change it to 3305 so i dont have to kill the mysql process or cahnge it to sometihng else

File: app/group/factories/group_factory.py [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    class Meta:
+        model = Group
+
+    slug = factory.Sequence(lambda n: "Group{0}".format(n))
+    description = factory.Faker("sentence", nb_words=100, variable_nb_words=True)
+    contact = factory.LazyAttributeSequence(
+        lambda o, n: "%s@s%d.example.com" % (o.slug, n)
+    )
+    created = 2019
+    type = type = factory.Iterator(GroupType)
```
Comment (Eirsteir) [inline app/group/factories/group_factory.py:19]: Beautiful! 😍

File: app/group/factories/group_factory.py [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+
+class GroupFactory(DjangoModelFactory):
+    class Meta:
+        model = Group
+
+    slug = factory.Sequence(lambda n: "Group{0}".format(n))
+    description = factory.Faker("sentence", nb_words=100, variable_nb_words=True)
+    contact = factory.LazyAttributeSequence(
+        lambda o, n: "%s@s%d.example.com" % (o.slug, n)
```
Comment (Eirsteir) [inline app/group/factories/group_factory.py:16]: You can use the newer f-string syntax here 😉

File: app/group/tests/test_group_intergration.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+def test_retrieve_as_anonymous_user(request_factory, group):
+
+    request = request_factory.get(f"/group/{group.slug}")
+    response = get_response(request, group)
+
+    assert response.status_code == 200
+
+
+@pytest.mark.django_db
+def test_retrieve_as_anonymous_user(request_factory, group):
```
Comment (Eirsteir) [inline app/group/tests/test_group_intergration.py:46]: This is the same test as above

File: app/group/tests/test_group_intergration.py [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    assert response.status_code == 403
+
+@pytest.mark.django_db
+@pytest.mark.parametrize(
+    ("group_name", "expected_status_code", "new_description"),
+    [
+        (AdminGroup.HS, 200, "New Description"),
+        (AdminGroup.DEVKOM, 200, "New Description"),
+        (AdminGroup.NOK, 403, "New Description"),
+        (AdminGroup.PROMO, 403, "New Description"),
```
Comment (Eirsteir) [inline app/group/tests/test_group_intergration.py:88]: Will the new description be 'New Description' if the status code was 403 Forbidden?

File: app/group/tests/test_group_intergration.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        f"/group/{group.slug}/", data=data, format="json", HTTP_X_CSRF_TOKEN=token.key
+    )
+    response = get_response(request, group, user)
+    group.refresh_from_db()
+
+    assert response.status_code == expected_status_code
+    assert group.description == expected_description
+
+@pytest.mark.django_db
+def test_create_makes_parent_group(request_factory, group, user):
```
Comment (Eirsteir) [inline app/group/tests/test_group_intergration.py:116]: Is this supposed to be named `test_create_makes_group_if_not_found`?

File: app/group/views/group.py [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+class GroupViewSet(viewsets.ModelViewSet):
+    serializer_class = GroupSerializer
+    permission_classes = [IsHS]
+    queryset = Group.objects.all()
+    lookup_field = "slug"
+    pagination_class = BasePagination
+
+    def get_permissions(self):
+        if self.request.method in ["GET"]:
+            self.permission_classes = [IsMember]
```
Comment (Eirsteir) [inline app/group/views/group.py:22]: You have written tests that contradict this behaviour. IsMember checks for a token in the request and denied access if it is not found. This means that an anonymous user should not be able to list or retrieve groups, yet the two first tests asserts that this is indeed possible. 

Please try running the tests and make sure this is not the case or correct the logic.

File: app/group/views/group.py [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+        except Group.DoesNotExist:
+            return Response({"detail": _("Group not found.")}, status=404)
+
+    def update(self, request, slug):
+        try:
+            group = Group.objects.get(slug=slug)
+            serializer = GroupSerializer(group, data=request.data, partial=True)
+            if serializer.is_valid():
+                serializer.save()
+                return Response({"detail": _("Group created")}, status=201)
```
Comment (Eirsteir) [inline app/group/views/group.py:41]: I had a recent chat with Olaf, who requested that we responded with the entity/object on both create and update. You can safely return the group in the response here 😊

File: app/group/tests/test_group_intergration.py [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+
+    data = {
+        "name": parent_name,
+        "slug": "group",
+    }
+
+    request = request_factory.put(f"/group/{group.slug}/", data=data)
+    response = get_response(request, group, user)
+    group.refresh_from_db()
+    assert group.name == parent_name
```
Comment (Eirsteir) [inline app/group/tests/test_group_intergration.py:128]: How about a test for create when a group already exists?

**Comment by Eirsteir** [comment]
The tests look great, really solid work here.

The only thing I am missing is docstrings. This makes it easier to follow along and ge a grasp of the expected behaviour is 😄 

Lastly, what do you think about placing all common code used in both apps in a new app called something like 'commons', ie common permissions (IsHS, IsMember etc), utils, basepagination, basemodel etc?

File: app/enums.py [hunk 20]
```diff
@@ -22,3 +22,12 @@ class AdminGroup(ChoiceEnum):
...
     PROMO = "Promo"
+
+
+class GroupType(ChoiceEnum):
+    THILDE = "THILDE"
+    STYRE = "Styre"
+    UNDERGRUPPE = "Undergruppe"
+    KOMMITE = "Kommite"
+    STUDIEAAR = "Studieår"
+    OTHER = "Annet"
```
Comment (Zenjjim) [inline app/enums.py:33]: Siden vi koder på engelsk, burde vel Enums også være på engelsk. f.eks. BOARD = "Styre", SUBGROUP="undergruppe"...

File: app/group/models/group.py [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+
+from app.util.enums import GroupType
+from app.util.models import OptionalImage
+
+
+class Group(DjangoGroup, OptionalImage):
+    slug = models.SlugField(max_length=50, primary_key=True)
+    description = models.TextField(max_length=1000, null=True, blank=True)
+    contact = models.EmailField(max_length=200, null=True, blank=True)
+    created = models.DateTimeField(null=True, blank=True)
```
Comment (Zenjjim) [inline app/group/models/group.py:14]: Om du importerer baseModel vil du få med created og updated

File: app/group/views/group.py [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+from ..models import Group
+from ..serializers import GroupSerializer
+
+
+class GroupViewSet(viewsets.ModelViewSet):
+    serializer_class = GroupSerializer
+    permission_classes = [IsHS]
+    queryset = Group.objects.all()
+    lookup_field = "slug"
+    pagination_class = BasePagination
```
Comment (Zenjjim) [inline app/group/views/group.py:18]: Ønsker vi pagination for henting av gruppene?

File: app/group/views/group.py (hunk 17 above)
Comment (Zenjjim) [reply app/group/views/group.py:22]: Det er vel ønskelig at man skal kunne hente alle grupper uansett, også er det ulike mengde data avhengig om du er anon, medlem, hs etc?

**Review by Zenjjim — CHANGES_REQUESTED** [review]

File: app/common/enums.py [hunk 23]
```diff
@@ -22,3 +22,12 @@ class AdminGroup(ChoiceEnum):
     DEVKOM = "DevKom"
     NOK = "Nok"
     PROMO = "Promo"
+
+
+class GroupType(ChoiceEnum):
+    THILDE = "THILDE"
+    BOARD = "Styre"
+    SUBGROUPE = "Undergruppe"
```
Comment (Eirsteir) [inline app/common/enums.py:30]: Typo in SUBGROUP

File: app/group/factories/group_factory.py [hunk 24]
```diff
@@ -0,0 +1,17 @@
...
+
+from ..models import Group
+
+
+class GroupFactory(DjangoModelFactory):
+    class Meta:
+        model = Group
+
+    name = factory.Sequence(lambda n: "Group{0}".format(n))
+    slug = factory.Sequence(lambda n: "Group{0}".format(n))
```
Comment (Eirsteir) [inline app/group/factories/group_factory.py:14]: Use f-strings here

File: app/group/models/group.py [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+from app.common.enums import GroupType
+from app.util.models import BaseModel, OptionalImage
+
+
+class Group(OptionalImage, BaseModel):
+    name = models.CharField(max_length=50)
+    permissions = models.ManyToManyField(Permission, related_name="permission")
+    slug = models.SlugField(max_length=50, primary_key=True)
+    description = models.TextField(max_length=1000, null=True, blank=True)
+    contact = models.EmailField(max_length=200, null=True, blank=True)
```
Comment (Eirsteir) [inline app/group/models/group.py:15]: Write the field name out in its entirety for cleaner code

File: app/group/models/group.py [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+
+from enumchoicefield import EnumChoiceField
+
+from app.common.enums import GroupType
+from app.util.models import BaseModel, OptionalImage
+
+
+class Group(OptionalImage, BaseModel):
+    name = models.CharField(max_length=50)
+    permissions = models.ManyToManyField(Permission, related_name="permission")
```
Comment (Eirsteir) [inline app/group/models/group.py:12]: The related name should also be plural for consistency and cleaner code from the caller

File: app/group/views/group.py [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+    def get_permissions(self):
+        if self.request.method in ["GET"]:
+            self.permission_classes = []
+        if self.request.method in ["POST"]:
+            self.permission_classes = [IsHS]
+        return super(GroupViewSet, self).get_permissions()
+
+    def retrieve(self, request, slug):
+        try:
+            group = Group.objects.get(slug=slug)
```
Comment (Eirsteir) [inline app/group/views/group.py:29]: I haven't tested it properly, but can you try to use `group = self.get_object()` and see if it does the job?

File: app/group/views/group.py [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+            return Response(
+                {"detail": _("Group not found.")}, status=status.HTTP_404_NOT_FOUND
+            )
+
+    def update(self, request, slug):
+        try:
+            print(request)
+            print(request.data)
+            print(slug)
+            group = Group.objects.get(slug=slug)
```
Comment (Eirsteir) [inline app/group/views/group.py:44]: Same here as above

File: app/tests/test_group_intergration.py [hunk 29]
```diff
@@ -0,0 +1,163 @@
...
+        "slug": group.slug,
+    }
+
+
+def _get_group_put_data(group):
+    return {**_get_group_post_data(group), "description": "New Description"}
+
+
+def add_user_to_group_with_name(user, group_name):
+    group = DjangoGroup.objects.create(name=group_name)
```
Comment (Eirsteir) [inline app/tests/test_group_intergration.py:28]: Can you add a comment here informing that it should be changed in the future to use our Group model?

File: app/tests/test_group_intergration.py [hunk 30]
```diff
@@ -0,0 +1,163 @@
...
+    user.groups.add(group)
+
+
+def get_response(request, group=None, user=None):
```
Comment (Eirsteir) [inline app/tests/test_group_intergration.py:32]: It would be preferable if we used APIClient instead. It provides a much cleaner and simpler implementation

File: app/group/models/group.py (hunk 25 above)
Comment (Eirsteir) [reply app/group/models/group.py:15]: Ie, contact_email 👍

File: app/group/views/group.py [hunk 31]
```diff
@@ -0,0 +1,66 @@
...
+        """Updates a spesific group by slug"""
+        try:
+            group = self.get_object()
+            serializer = GroupSerializer(group, data=request.data, partial=True)
+            if serializer.is_valid():
+                serializer.save()
+                return Response(data=serializer.data, status=status.HTTP_200_OK)
+        except Group.DoesNotExist:
+            return Response(
+                {"detail": ("Gruppe eksisterer ikke")}, status=status.HTTP_404_NOT_FOUND
```
Comment (Zenjjim) [inline app/group/views/group.py:47]: gruppen*

File: app/group/models/group.py [hunk 32]
```diff
@@ -0,0 +1,26 @@
...
+from django.db import models
+
+from enumchoicefield import EnumChoiceField
+
+from app.common.enums import GroupType
+from app.util.models import BaseModel, OptionalImage
+
+
+class Group(OptionalImage, BaseModel):
+    """Model for Costum Groups"""
```
Comment (Zenjjim) [inline app/group/models/group.py:11]: custom*

File: app/group/factories/group_factory.py [hunk 33]
```diff
@@ -0,0 +1,21 @@
...
+
+class GroupFactory(DjangoModelFactory):
+    """Factory that creates a generic group"""
+
+    class Meta:
+        model = Group
+
+    name = factory.Sequence(lambda n: f"Group{n}")
+    slug = factory.Sequence(lambda n: f"Group{n}")
+    group_description = factory.Faker("sentence", nb_words=100, variable_nb_words=True)
```
Comment (Zenjjim) [inline app/group/factories/group_factory.py:17]: drop group_

File: app/group/models/group.py [hunk 34]
```diff
@@ -0,0 +1,26 @@
...
+
+    name = models.CharField(max_length=50)
+    group_permissions = models.ManyToManyField(
```
Comment (Zenjjim) [inline app/group/models/group.py:14]: -group_

File: app/group/serializers/group.py [hunk 35]
```diff
@@ -0,0 +1,20 @@
+from rest_framework import serializers
+
+from ..models import Group
```
Comment (Zenjjim) [inline app/group/serializers/group.py:3]: absolute import

File: app/group/urls.py [hunk 36]
```diff
@@ -0,0 +1,10 @@
+from django.conf.urls import include, url
+from rest_framework import routers
+
+from .views import GroupViewSet
```
Comment (Zenjjim) [inline app/group/urls.py:4]: abs imp

File: app/group/views/group.py [hunk 37]
```diff
@@ -0,0 +1,66 @@
...
+class GroupViewSet(viewsets.ModelViewSet):
+    """API endpoint for Groups"""
+
+    serializer_class = GroupSerializer
+    permission_classes = [IsNoKorPromo]
+    queryset = Group.objects.all()
+    lookup_field = "slug"
+
+    def get_permissions(self):
+        if self.request.method in ["GET"]:
```
Comment (Zenjjim) [inline app/group/views/group.py:18]: self.action.view

File: app/group/models/group.py [hunk 38]
```diff
@@ -8,16 +8,16 @@
 
 
 class Group(OptionalImage, BaseModel):
-    """Model for Costum Groups"""
+    """Model for Custom Groups"""
 
     name = models.CharField(max_length=50)
-    group_permissions = models.ManyToManyField(
+    permissions = models.ManyToManyField(
         Permission, related_name="auth_permissions", blank=True
```
Comment (Eirsteir) [inline app/group/models/group.py:15]: The related name should really be `groups` since we are defining how we should access the groups from the Permission side of the relationship. In the same way in Registration, we have `event = ForeignKey(..., related_name="registrations")`

**Review by Eirsteir — APPROVED** [review]
Really fine work you have done here 😄 I especially appreciate the effort you put into writing tests and polishing the code

**Review by Zenjjim — APPROVED** [review]
